File: backend/routes/teacher.py
from flask import Blueprint, request, jsonify, current_app
from models import db, Student, Teacher, Attendance, ClassSession, Subject, AuditLog
from auth import role_required
from datetime import datetime, timezone
import secrets

teacher_bp = Blueprint('teacher', __name__)


@teacher_bp.route('/api/teacher/start-session', methods=['POST'])
@role_required('teacher')
def start_session():
    """Start a new class session."""
    data = request.get_json()
    if not data or not data.get('subject_id'):
        return jsonify({'error': 'Subject ID required'}), 400

    teacher = request.current_user.teacher
    if not teacher:
        return jsonify({'error': 'Teacher profile not found'}), 404

    subject = Subject.query.get(data['subject_id'])
    if not subject:
        return jsonify({'error': 'Subject not found'}), 404

    # Check for existing active session
    active = ClassSession.query.filter_by(
        teacher_id=teacher.id, is_active=True
    ).first()
    if active:
        return jsonify({'error': 'You already have an active session', 'session': active.to_dict()}), 409

    session_obj = ClassSession(
        session_token=secrets.token_urlsafe(32),
        subject_id=subject.id,
        teacher_id=teacher.id,
        department=subject.department,
        semester=subject.semester,
        start_time=datetime.now(timezone.utc),
    )
    db.session.add(session_obj)

    audit = AuditLog(user_id=request.current_user.id, action='start_session',
                     details=f'Session started for {subject.code}',
                     ip_address=request.remote_addr)
    db.session.add(audit)
    db.session.commit()

    return jsonify({'message': 'Session started', 'session': session_obj.to_dict()}), 201


@teacher_bp.route('/api/teacher/end-session/<int:session_id>', methods=['POST'])
@role_required('teacher')
def end_session(session_id):
    """End an active class session."""
    teacher = request.current_user.teacher
    session_obj = ClassSession.query.filter_by(
        id=session_id, teacher_id=teacher.id, is_active=True
    ).first()

    if not session_obj:
        return jsonify({'error': 'Active session not found'}), 404

    session_obj.is_active = False
    session_obj.end_time = datetime.now(timezone.utc)

    # Mark all students who weren't scanned as absent
    enrolled_students = Student.query.filter_by(
        department=session_obj.department,
        semester=session_obj.semester,
    ).all()

    present_ids = {a.student_id for a in
                   Attendance.query.filter_by(session_id=session_obj.id).all()}

    for student in enrolled_students:
        if student.id not in present_ids:
            absent_record = Attendance(
                student_id=student.id,
                session_id=session_obj.id,
                subject_id=session_obj.subject_id,
                status='absent',
                marked_by='auto',
            )
            db.session.add(absent_record)

    audit = AuditLog(user_id=request.current_user.id, action='end_session',
                     details=f'Session {session_id} ended',
                     ip_address=request.remote_addr)
    db.session.add(audit)
    db.session.commit()

    return jsonify({'message': 'Session ended', 'session': session_obj.to_dict()}), 200


@teacher_bp.route('/api/teacher/scan-qr', methods=['POST'])
@role_required('teacher')
def scan_qr():
    """Verify scanned QR code and mark attendance."""
    data = request.get_json()
    if not data or not data.get('qr_token') or not data.get('session_id'):
        return jsonify({'error': 'QR token and session ID required'}), 400

    teacher = request.current_user.teacher
    session_obj = ClassSession.query.filter_by(
        id=data['session_id'], teacher_id=teacher.id, is_active=True
    ).first()

    if not session_obj:
        return jsonify({'error': 'No active session found'}), 404

    qr_service = current_app.config.get('QR_SERVICE')
    result = qr_service.verify_qr_token(data['qr_token'])

    if not result['valid']:
        return jsonify({'error': result['error']}), 400

    student = Student.query.get(result['student_id'])
    if not student:
        return jsonify({'error': 'Student not found'}), 404

    # Verify student belongs to this class
    if student.department != session_obj.department or student.semester != session_obj.semester:
        return jsonify({'error': 'Student not enrolled in this class'}), 403

    # Check if already marked
    existing = Attendance.query.filter_by(
        student_id=student.id, session_id=session_obj.id
    ).first()

    if existing and existing.status == 'present':
        return jsonify({'message': 'Already marked present', 'attendance': existing.to_dict()}), 200

    if existing:
        existing.status = 'present'
        existing.marked_at = datetime.now(timezone.utc)
        existing.marked_by = 'qr_scan'
    else:
        attendance = Attendance(
            student_id=student.id,
            session_id=session_obj.id,
            subject_id=session_obj.subject_id,
            status='present',
            marked_by='qr_scan',
            device_info=data.get('device_info'),
        )
        db.session.add(attendance)
        existing = attendance

    audit = AuditLog(user_id=request.current_user.id, action='mark_attendance',
                     details=f'QR scan: {student.roll_number} marked present',
                     ip_address=request.remote_addr)
    db.session.add(audit)
    db.session.commit()

    return jsonify({
        'message': f'{student.user.full_name} marked present',
        'attendance': existing.to_dict(),
        'student': student.to_dict(),
    }), 200


@teacher_bp.route('/api/teacher/session-attendance/<int:session_id>', methods=['GET'])
@role_required('teacher')
def get_session_attendance(session_id):
    """Get attendance list for a session."""
    teacher = request.current_user.teacher
    session_obj = ClassSession.query.filter_by(
        id=session_id, teacher_id=teacher.id
    ).first()

    if not session_obj:
        return jsonify({'error': 'Session not found'}), 404

    # Get all students for this department/semester
    students = Student.query.filter_by(
        department=session_obj.department,
        semester=session_obj.semester,
    ).all()

    attendance_map = {}
    for a in Attendance.query.filter_by(session_id=session_obj.id).all():
        attendance_map[a.student_id] = a.to_dict()

    student_list = []
    for s in students:
        att = attendance_map.get(s.id)
        student_list.append({
            'student': s.to_dict(),
            'status': att['status'] if att else 'not_marked',
            'marked_at': att['marked_at'] if att else None,
        })

    present_count = sum(1 for s in student_list if s['status'] == 'present')
    total = len(student_list)

    return jsonify({
        'session': session_obj.to_dict(),
        'students': student_list,
        'present_count': present_count,
        'total_students': total,
    }), 200


@teacher_bp.route('/api/teacher/subjects', methods=['GET'])
@role_required('teacher')
def get_teacher_subjects():
    """Get subjects assigned to the teacher."""
    teacher = request.current_user.teacher
    if not teacher:
        return jsonify({'error': 'Teacher profile not found'}), 404

    subjects = Subject.query.filter_by(teacher_id=teacher.id).all()
    return jsonify({'subjects': [s.to_dict() for s in subjects]}), 200


@teacher_bp.route('/api/teacher/sessions', methods=['GET'])
@role_required('teacher')
def get_teacher_sessions():
    """Get all sessions for the teacher."""
    teacher = request.current_user.teacher
    if not teacher:
        return jsonify({'error': 'Teacher profile not found'}), 404

    sessions = ClassSession.query.filter_by(teacher_id=teacher.id)\
        .order_by(ClassSession.created_at.desc()).limit(50).all()
    return jsonify({'sessions': [s.to_dict() for s in sessions]}), 200


@teacher_bp.route('/api/teacher/manual-attendance', methods=['POST'])
@role_required('teacher')
def manual_attendance():
    """Manually mark attendance (for technical issues)."""
    data = request.get_json()
    if not data or not data.get('student_id') or not data.get('session_id') or not data.get('status'):
        return jsonify({'error': 'student_id, session_id, and status required'}), 400

    teacher = request.current_user.teacher
    session_obj = ClassSession.query.filter_by(
        id=data['session_id'], teacher_id=teacher.id
    ).first()

    if not session_obj:
        return jsonify({'error': 'Session not found'}), 404

    student = Student.query.get(data['student_id'])
    if not student:
        return jsonify({'error': 'Student not found'}), 404

    if data['status'] not in ('present', 'absent', 'late'):
        return jsonify({'error': 'Invalid status'}), 400

    existing = Attendance.query.filter_by(
        student_id=student.id, session_id=session_obj.id
    ).first()

    if existing:
        existing.status = data['status']
        existing.marked_by = 'manual_teacher'
        existing.marked_at = datetime.now(timezone.utc)
    else:
        attendance = Attendance(
            student_id=student.id,
            session_id=session_obj.id,
            subject_id=session_obj.subject_id,
            status=data['status'],
            marked_by='manual_teacher',
        )
        db.session.add(attendance)

    audit = AuditLog(user_id=request.current_user.id, action='manual_attendance',
                     details=f'Manual: {student.roll_number} marked {data["status"]}',
                     ip_address=request.remote_addr)
    db.session.add(audit)
    db.session.commit()

    return jsonify({'message': f'Attendance updated for {student.user.full_name}'}), 200

File: backend/routes/teacher.js
const express = require('express');
const { verifyToken, requireRole } = require('../middleware/auth');
const { Teacher, Student, Attendance, Subject } = require('../models');
const QRService = require('../services/QRService');
const router = express.Router();

/**
 * GET /api/teacher/dashboard
 * Get teacher dashboard data
 */
router.get('/dashboard', verifyToken, requireRole('teacher'), async (req, res) => {
  try {
    const teacher = await Teacher.findOne({ userId: req.userId });
    if (!teacher) {
      return res.status(404).json({ error: 'Teacher profile not found' });
    }

    // Get attendance statistics
    const totalAttendanceRecords = await Attendance.countDocuments({ teacherId: teacher._id });
    const todayAttendance = await Attendance.countDocuments({
      teacherId: teacher._id,
      scannedAt: {
        $gte: new Date().setHours(0, 0, 0, 0),
        $lt: new Date().setHours(23, 59, 59, 999),
      },
    });

    res.json({
      success: true,
      teacher: teacher.toObject(),
      statistics: {
        totalAttendanceRecords,
        todayAttendance,
      },
    });
  } catch (error) {
    console.error('Error fetching dashboard:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * GET /api/teacher/subjects
 * Get subjects assigned to the teacher
 */
router.get('/subjects', verifyToken, requireRole('teacher'), async (req, res) => {
  try {
    const teacher = await Teacher.findOne({ userId: req.userId });
    if (!teacher) {
      return res.status(404).json({ error: 'Teacher profile not found' });
    }

    const subjects = await Subject.find({
      $or: [
        { teacherId: teacher._id },
        { assignedTeachers: teacher._id },
      ],
      isActive: true,
    }).sort({ subjectCode: 1 });

    res.json({
      success: true,
      subjects: subjects.map(s => ({
        id: s._id,
        code: s.subjectCode,
        name: s.subjectName,
        department: s.department,
        semester: s.semester,
      })),
    });
  } catch (error) {
    console.error('Error fetching subjects:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * POST /api/teacher/start-session
 * Start a new attendance session for a subject
 */
router.post('/start-session', verifyToken, requireRole('teacher'), async (req, res) => {
  try {
    const { subjectId } = req.body;

    if (!subjectId) {
      return res.status(400).json({ error: 'Subject ID is required' });
    }

    const teacher = await Teacher.findOne({ userId: req.userId });
    if (!teacher) {
      return res.status(404).json({ error: 'Teacher profile not found' });
    }

    const subject = await Subject.findById(subjectId);
    if (!subject) {
      return res.status(404).json({ error: 'Subject not found' });
    }

    // Create a session ID based on timestamp
    const sessionId = `session_${Date.now()}`;

    res.json({
      success: true,
      session: {
        id: sessionId,
        subjectId: subject._id,
        subject_code: subject.subjectCode,
        subject_name: subject.subjectName,
        is_active: true,
        started_at: new Date().toISOString(),
      },
    });
  } catch (error) {
    console.error('Error starting session:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * GET /api/teacher/attendance-records
 * Get attendance records created by teacher
 */
router.get('/attendance-records', verifyToken, requireRole('teacher'), async (req, res) => {
  try {
    const teacher = await Teacher.findOne({ userId: req.userId });
    if (!teacher) {
      return res.status(404).json({ error: 'Teacher profile not found' });
    }

    const { page = 1, limit = 20, month, year } = req.query;

    let filter = { teacherId: teacher._id };

    // Filter by month and year if provided
    if (month && year) {
      const startDate = new Date(year, month - 1, 1);
      const endDate = new Date(year, month, 0);
      filter.scannedAt = { $gte: startDate, $lte: endDate };
    }

    const skip = (page - 1) * limit;
    const records = await Attendance.find(filter)
      .populate('studentId')
      .sort({ scannedAt: -1 })
      .skip(skip)
      .limit(parseInt(limit));

    const total = await Attendance.countDocuments(filter);

    res.json({
      success: true,
      records,
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total,
        pages: Math.ceil(total / limit),
      },
    });
  } catch (error) {
    console.error('Error fetching attendance records:', error);
    res.status(500).json({ error: error.message });
  }
});



/**
 * POST /api/teacher/scan-student-qr
 * Teacher scans student's personal QR code to mark attendance
 */
router.post('/scan-student-qr', verifyToken, requireRole('teacher'), async (req, res) => {
  try {
    const { qrData } = req.body;

    console.log('📨 Received QR Scan Request');
    console.log('QR Data:', qrData);

    if (!qrData) {
      console.error('❌ No QR data provided');
      return res.status(400).json({ error: 'QR data is required' });
    }

    const teacher = await Teacher.findOne({ userId: req.userId });
    if (!teacher) {
      console.error('❌ Teacher profile not found for user:', req.userId);
      return res.status(404).json({ error: 'Teacher profile not found' });
    }

    // Decode student QR data
    let decodedData;
    try {
      console.log('🔍 Attempting to decode QR data...');
      decodedData = QRService.decodeStudentQRData(qrData);
      console.log('✅ Decoded data:', decodedData);
    } catch (error) {
      console.error('❌ QR decode error:', error.message);
      return res.status(400).json({ error: 'Invalid QR code format: ' + error.message });
    }

    // Validate QR data contains required fields
    if (!decodedData.enrollmentNo || !decodedData.studentId) {
      console.error('❌ Invalid QR data - missing required fields:', decodedData);
      return res.status(400).json({ error: 'Invalid student QR code data. Missing enrollmentNo or studentId' });
    }

    // Find student
    const student = await Student.findById(decodedData.studentId).populate('userId');
    if (!student) {
      console.error('❌ Student not found with ID:', decodedData.studentId);
      return res.status(404).json({ error: 'Student not found' });
    }

    console.log('✅ Student found:', student.enrollmentNo, student.userId?.fullName);

    // Check if attendance already marked in the last 2 minutes (prevent duplicate scans)
    const existingAttendance = await Attendance.findOne({
      studentId: student._id,
      teacherId: teacher._id,
      scannedAt: {
        $gte: new Date(Date.now() - 2 * 60 * 1000), // Within last 2 minutes
      },
    });

    if (existingAttendance) {
      console.log('⚠️ Duplicate scan - attendance already marked in last 2 minutes');
      return res.status(409).json({
        error: 'Attendance already marked for this student. Please wait before rescanning.',
        student: {
          id: student._id,
          enrollmentNo: student.enrollmentNo,
          name: student.userId?.fullName || student.enrollmentNo,
        },
      });
    }

    // Create attendance record
    const attendance = new Attendance({
      studentId: student._id,
      enrollmentNo: student.enrollmentNo,
      teacherId: teacher._id,
      qrCodeHash: `student_${decodedData.enrollmentNo}`,
      qrGeneratedAt: new Date(),
      ipAddress: req.ip,
      deviceInfo: req.headers['user-agent'],
      attendanceStatus: 'present',
      scannedAt: new Date(),
    });

    await attendance.save();

    const studentName = student.userId?.fullName || student.enrollmentNo;

    console.log('✅ Attendance marked successfully for:', studentName);

    res.json({
      success: true,
      message: `✅ ${studentName} - Marked Present`,
      attendance: attendance.toObject(),
      student: {
        id: student._id,
        enrollmentNo: student.enrollmentNo,
        name: studentName,
        department: student.department,
      },
    });
  } catch (error) {
    console.error('❌ Error scanning student QR:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * GET /api/teacher/session-attendance/:sessionId
 * Get today's attendance for a session with optional subject filter
 */
router.get('/session-attendance/:sessionId', verifyToken, requireRole('teacher'), async (req, res) => {
  try {
    console.log('📋 Fetching session attendance for sessionId:', req.params.sessionId);

    const teacher = await Teacher.findOne({ userId: req.userId });
    if (!teacher) {
      console.error('❌ Teacher profile not found');
      return res.status(404).json({ error: 'Teacher profile not found' });
    }

    const { subjectId } = req.query;
    console.log('Subject filter:', subjectId);

    // Get all students
    const allStudents = await Student.find({}).populate('userId');
    console.log('Total students found:', allStudents.length);
    
    // Get today's attendance marked by this teacher
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    const tomorrow = new Date(today);
    tomorrow.setDate(tomorrow.getDate() + 1);

    console.log('Checking attendance between:', today, 'and', tomorrow);

    const todayRecords = await Attendance.find({
      teacherId: teacher._id,
      scannedAt: {
        $gte: today,
        $lt: tomorrow,
      },
    }).populate('studentId');

    console.log('Attendance records found today:', todayRecords.length);
    todayRecords.forEach(record => {
      console.log('  -', record.studentId?.enrollmentNo, record.attendanceStatus);
    });

    const presentStudentIds = new Set(todayRecords.map(r => String(r.studentId?._id)));

    // Build student list with attendance status
    const studentList = allStudents.map(student => ({
      id: student._id,
      enrollmentNo: student.enrollmentNo,
      full_name: student.userId?.fullName || student.enrollmentNo,
      roll_number: student.rollNumber || student.enrollmentNo,
      department: student.department,
      status: presentStudentIds.has(String(student._id)) ? 'present' : 'absent',
    }));

    const presentCount = studentList.filter(s => s.status === 'present').length;

    console.log('✅ Attendance list prepared:', presentCount, 'present out of', studentList.length);

    res.json({
      success: true,
      students: studentList,
      present_count: presentCount,
      total_students: studentList.length,
    });
  } catch (error) {
    console.error('❌ Error fetching session attendance:', error);
    res.status(500).json({ error: error.message });
  }
});



/**
 * POST /api/teacher/mark-attendance-manual
 * Manually mark student attendance (teacher action)
 */
router.post('/mark-attendance-manual', verifyToken, requireRole('teacher'), async (req, res) => {
  try {
    const { studentId, sessionId, status } = req.body;

    if (!studentId || !sessionId || !status) {
      return res.status(400).json({ error: 'Student ID, session ID, and status are required' });
    }

    if (!['present', 'absent', 'late', 'excused'].includes(status)) {
      return res.status(400).json({ error: 'Invalid attendance status' });
    }

    const teacher = await Teacher.findOne({ userId: req.userId });
    if (!teacher) {
      return res.status(404).json({ error: 'Teacher profile not found' });
    }

    const student = await Student.findById(studentId).populate('userId');
    if (!student) {
      return res.status(404).json({ error: 'Student not found' });
    }

    // Check if attendance already marked in the last 5 minutes
    const existingAttendance = await Attendance.findOne({
      studentId: student._id,
      teacherId: teacher._id,
      scannedAt: {
        $gte: new Date(Date.now() - 5 * 60 * 1000),
      },
    });

    if (existingAttendance) {
      // Update existing record instead of creating duplicate
      existingAttendance.attendanceStatus = status;
      existingAttendance.remarks = 'Manually marked by teacher';
      await existingAttendance.save();

      return res.json({
        success: true,
        message: `Attendance updated for ${student.userId?.fullName || student.enrollmentNo}`,
        attendance: existingAttendance.toObject(),
      });
    }

    // Create new attendance record
    const attendance = new Attendance({
      studentId: student._id,
      enrollmentNo: student.enrollmentNo,
      teacherId: teacher._id,
      qrCodeHash: sessionId,
      qrGeneratedAt: new Date(),
      ipAddress: req.ip,
      deviceInfo: req.headers['user-agent'],
      attendanceStatus: status,
      remarks: 'Manually marked by teacher',
      scannedAt: new Date(),
    });

    await attendance.save();

    res.json({
      success: true,
      message: `✅ Attendance marked for ${student.userId?.fullName || student.enrollmentNo}`,
      attendance: attendance.toObject(),
    });
  } catch (error) {
    console.error('Error marking manual attendance:', error);
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;
